src/client_builder.rs

```diff
@@ -61,41 +155,132 @@ where
+        self.no_tls_verify = no_tls_verify;
+        self
+    }
+
+    /// Makes a [`Client`] connection using the native TLS backend.
+    ///
+    /// Forces TLS unless [`ClientBuilder::mode(ConnectionMode::StartTls)`] was called
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// use imap::TlsKind;
+    /// let client = ClientBuilder::new("imap.example.com", 993).native_tls()?;
+    /// # Ok(())
+    /// # }
+    /// ```
     #[cfg(feature = "native-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "native-tls")))]
-    pub fn native_tls(&mut self) -> Result<Client<TlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = TlsConnector::builder().build()?;
-            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
-        })
+    #[deprecated = "Use tls_kind(TlsKind::Native).connect() instead"]
+    pub fn native_tls(&mut self) -> Result<Client<Connection>> {
+        if self.mode == ConnectionMode::Tcp {
+            self.mode(ConnectionMode::Tls);
+        }
+
+        self.tls_kind(TlsKind::Native).connect()
     }
 
-    /// Return a new [`Client`] using `rustls` transport.
+    /// Makes a [`Client`] connection using the rustls TLS backend.
+    ///
+    /// Forces TLS unless [`ClientBuilder::starttls()`] was called
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// let client = ClientBuilder::new("imap.example.com", 993).rustls()?;
+    /// # Ok(())
+    /// # }
+    /// ```
     #[cfg(feature = "rustls-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-tls")))]
-    pub fn rustls(&mut self) -> Result<Client<RustlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = RustlsConnector::new_with_native_certs()?;
-            Ok(ssl_conn.connect(domain, tcp)?)
+    #[deprecated = "Use tls_kind(TlsKind::Rust).connect() instead"]
+    pub fn rustls(&mut self) -> Result<Client<Connection>> {
+        if self.mode == ConnectionMode::Tcp {
+            self.mode(ConnectionMode::Tls);
+        }
+
+        self.tls_kind(TlsKind::Rust).connect()
+    }
+
+    /// Make a [`Client`] using the configuration.
+    ///
+    /// ```no_run
+    /// # use imap::ClientBuilder;
+    /// # {} #[cfg(feature = "rustls-tls")]
+    /// # fn main() -> Result<(), imap::Error> {
+    /// let client = ClientBuilder::new("imap.example.com", 143)
+    ///     .starttls().connect()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn connect(&self) -> Result<Client<Connection>> {
+        self.connect_with(|_domain, tcp| {
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            return self.build_tls_connection(tcp);
+            #[cfg(all(not(feature = "native-tls"), not(feature = "rustls-tls")))]
+            return Ok(tcp);
```

Comment:
This is a nit, but I'd kind of like for `handshake` to be called something like `tls_handshake`, and then for us to ensure in `connect_with` that we never call it if we don't have any TLS capability features enabled. Which would then make this be a `unreachable!` rather than `return Ok(tcp)`. What do you think?